### PR TITLE
enable the DF bit on macOS

### PR DESCRIPTION
**Pull request by sukunrt**
Closes: #3793

### CODE REVIEW

File: sys_conn_df_darwin.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+	"github.com/quic-go/quic-go/internal/utils"
+)
+
+func setDF(rawConn syscall.RawConn) (bool, error) {
+	uname := &unix.Utsname{}
+	err := unix.Uname(uname)
+	if err != nil {
+		return false, err
+	}
+	if !isAtLeastOS11(uname) {
```
Comment (marten-seemann) [inline sys_conn_df_darwin.go:22]: Can you add a comment here why we're testing for this?

File: sys_conn_df_darwin.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+		return false, errors.New("setting DF failed for both IPv4 and IPv6")
+	}
+	return true, nil
+}
+
+func isMsgSizeErr(err error) bool {
+	return errors.Is(err, unix.EMSGSIZE)
+}
+
+func isAtLeastOS11(uname *unix.Utsname) bool {
```
Comment (marten-seemann) [inline sys_conn_df_darwin.go:55]:
```suggestion
func isAtLeastMacOS11(uname *unix.Utsname) bool {
```

File: sys_conn_df_darwin.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+	"github.com/quic-go/quic-go/internal/utils"
+)
+
+func setDF(rawConn syscall.RawConn) (bool, error) {
+	uname := &unix.Utsname{}
+	err := unix.Uname(uname)
+	if err != nil {
+		return false, err
+	}
```
Comment (marten-seemann) [inline sys_conn_df_darwin.go:21]: Nit: Should these lines be moved into the `isAtLeast` function?

File: sys_conn_df_darwin.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+	switch {
+	case errDFIPv4 == nil && errDFIPv6 == nil:
+		utils.DefaultLogger.Debugf("Setting DF for IPv4 and IPv6.")
+	case errDFIPv4 == nil && errDFIPv6 != nil:
+		utils.DefaultLogger.Debugf("Setting DF for IPv4.")
+	case errDFIPv4 != nil && errDFIPv6 == nil:
+		utils.DefaultLogger.Debugf("Setting DF for IPv6.")
+		// On macOS DF bit for IPv4 cannot be set on dual-stack listeners.
```
Comment (marten-seemann) [inline sys_conn_df_darwin.go:42]:
```suggestion
		// On macOS, the syscall for setting DF bit for IPv4 fails on dual-stack listeners.
		// Treat the connection as not having DF enabled, even though the DF bit will be set 
		// when used for IPv6.
		// See https://github.com/quic-go/quic-go/issues/3793 for details.
```

**Comment by marten-seemann** [comment]
Looking at this logic and comparing it to what we do for Linux, I'm wondering why we have these two cases there:
https://github.com/quic-go/quic-go/blob/0fe21c7d6fc29a1f493a751ea524c3c9428f9848/sys_conn_df_linux.go#L31-L34

I assume that they occur for v4-only and v6-only sockets only, but that's just an assumption. It would be nice if we had a way to test if a socket is v4-only (or v6-only), and only do the correct syscall then.

Obviously, this is out of scope for this PR, but might be a nice follow-up.

File: sys_conn_df_darwin.go (hunk 1 above)
Comment (sukunrt) [reply sys_conn_df_darwin.go]: I picked this up from chromium source code and also some random answer on stackoverflow / github. I couldn't find a more official source so I've just added a comment saying it's only supported from macOS 11.

Comment (marten-seemann) [reply sys_conn_df_darwin.go]: Could you link the Chromium source code and / or the Stackoverflow here?

Comment (sukunrt) [reply sys_conn_df_darwin.go]: done

**Review by marten-seemann — APPROVED** [review]
Thank you @sukunrt!

**Pull request merged**